proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,553 @@
+allows the mechanism to handle not just backtraces but any arbitrary
+context-relevant value which may be inspected by a handler for analysis and logging.
+To avoid breaking existing users of ``SomeException``, we introduce this
+context as an implicit parameter constraint:  ::
+
+    data SomeException where
+        SomeException :: forall a. (Exception a, ?exceptionContext :: ExceptionContext)
+                      => a -> SomeException
+
+    data ExceptionContext = ExceptionContext [SomeExceptionAnnotation]
+
+    -- | Contexts are merged by concatenation
+    instance Semigroup ExceptionContext
+    instance Monoid ExceptionContext
+
+Following the example of the ``Exception`` class, we propose that exception
+annotations implement a minimal typeclass which provides ``Typeable`` evidence
+and the ability to render the annotation to a ``String``: ::
+
+    data SomeExceptionAnnotation where
+        SomeExceptionAnnotation :: forall a. (ExceptionAnnotation a)
+                                => a -> SomeExceptionAnnotation
+
+    class Typeable a => ExceptionAnnotation a where
+        displayExceptionAnnotation :: a -> String
+        
+        default displayExceptionAnnotation :: (Show a) => a -> String
+        displayExceptionAnnotation = show
+
+To allow users to populate this new annotation field we propose that the
+``Exception`` class be extended with a new ``toExceptionWithContext`` method: ::
+
+    class (Typeable e, Show e) => Exception e where
+        -- These are unchanged:
+        toException            :: e -> SomeException
+        fromException          :: SomeException -> Maybe e
+        displayException       :: e -> String
+
+        -- This is new:
+        toExceptionWithContext :: e -> ExceptionContext -> SomeException
+        toExceptionWithContext e ?exceptionContext = SomeException e
+
+        -- toException is implemented in terms of toExceptionWithContext
+        toException e = toExceptionWithContext e mempty
+
+The ``Exception`` instance for ``SomeException`` would accumulate and display
+contexts: ::
+
+    instance Exception SomeException where
+        -- This is unchanged:
+        toException se = se
+        fromException = Just
+
+        -- toExceptionWithContext *adds* context to an existing SomeException:
+        toExceptionWithContext se ctxt = addExceptionContext ctxt se
+
+        -- displayException shows context after the exception itself:
+        displayException (SomeException e) =
+            displayException e ++ "\n" ++ displayExceptionContext ?exceptionContext
+
+    displayExceptionContext :: ExceptionContext -> String
+    displayExceptionContext = ...
+
+Since the ``SomeException``'s ``displayException`` implementation is used to
+by GHC's top-level exception handler to display uncaught exceptions, this
+change carries the consequence that uncaught exceptions will have their context
+automatically printed as part of the error message presented to the user.
```

Comment:
Perhaps add a forward reference to the section "Teach top-level handler to use displayException"?